### PR TITLE
Sync with dev

### PR DESCRIPTION
**Pull request by shramee**
* chore: bump to alpha.7

* chore:fmt

### CODE REVIEW

**Comment by enitrat** [comment]
just rebased, waiting for CI and then let's bring this to main

**Comment by shramee** [comment]
@enitrat Ready for review and merge!

File: exercises/modules/modules1.cairo [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#[test]
+fn test_mod_fn() {
+    // Fix this line to call take_order function from module
+    let order_result = take_order();
+
+    assert(order_result == 'order_taken', 'Order not taken');
+}
+
+
+mod tests {
```
Comment (enitrat) [inline exercises/modules/modules1.cairo:22]: annotate module with #[cfg(test)] (good practice)

File: exercises/modules/modules2.cairo [hunk 2]
```diff
@@ -0,0 +1,56 @@
+// I AM NOT DONE
+// These modules have some issues, can you fix?
+// Run `starklings hint modules2` or `hint` watch command for a hint.
+
+use debug::PrintTrait;
+const YEAR: felt252 = 2050;
```
Comment (enitrat) [inline exercises/modules/modules2.cairo:6]: let's avoid using felt252 where possible - we want to push good practices, a year fits in a u16 so we should use a u16 here

File: info.toml [hunk 3]
```diff
@@ -503,6 +503,23 @@ If you're having trouble updating the distance value in the `Fish` and `Dog` imp
...
+path = "exercises/modules/modules1.cairo"
+mode = "test"
+hint = """"""
+
+[[exercises]]
+name = "modules2"
+path = "exercises/modules/modules2.cairo"
+mode = "test"
+hint = """
+While using functions/structs etc from outside the module,
```
Comment (enitrat) [inline info.toml:519]: > functions/structs etc

etc: replace by and other items

File: info.toml [hunk 4]
```diff
@@ -503,6 +503,23 @@ If you're having trouble updating the distance value in the `Fish` and `Dog` imp
...
+you can use full paths or `use` keyword to create shortcuts.
```
Comment (enitrat) [inline info.toml:520]: > you can use full paths or `use` keyword to create shortcuts.

proposition: you can refer to them with their full path or import them in the current context with the `use` keyword;

File: exercises/modules/readme.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
+# Modules
+
+Modules allow creating individual scopes and organise your code better. Read about [modules in cairo book](https://cairo-book.github.io/ch06-02-defining-modules-to-control-scope.html).
+
+Here's some code to show modules at play,
+
+## Declaring modules
+
+Modules can be declared in two way,
```
Comment (enitrat) [inline exercises/modules/readme.md:9]: in two ways

File: exercises/modules/readme.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
+# Modules
+
+Modules allow creating individual scopes and organise your code better. Read about [modules in cairo book](https://cairo-book.github.io/ch06-02-defining-modules-to-control-scope.html).
+
+Here's some code to show modules at play,
+
+## Declaring modules
```
Comment (enitrat) [inline exercises/modules/readme.md:7]: I'm actually wondering if we should give this explanation and not just redirect to the book. 
The more content we re-write, the more we have to maintain, and the more errors can slip in.

This part feels like a duplicate, it goes too much in-depth to be a high-order summary, and not enough to really understand everything

File: exercises/modules/modules2.cairo [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            name,
+            year: YEAR,
+            made_by_phone,
+            made_by_email: ! made_by_phone,
+            item,
+        }
+    }
+}
+
+mod order_utils {
```
Comment (enitrat) [inline exercises/modules/modules2.cairo:29]: Perhaps a `TODO: explain what you have to do in this file` would be nice to have

File: exercises/starknet/README.md [hunk 8]
```diff
@@ -1,7 +1,87 @@
 # Starknet Smart Contracts
 
-Cairo is the language used to develop Smart Contracts on Starknet.
+A Starknet contract is pretty much a Cairo module, or rather, it's represented by one.
```
Comment (enitrat) [inline exercises/starknet/README.md:3]: > A Starknet contract is pretty much a Cairo module, or rather, it's represented by one.

A Starknet contract can be created by annotating a Cairo module with the `#[contract]` attribute

File: exercises/starknet/README.md [hunk 9]
```diff
@@ -1,7 +1,87 @@
...
+
+Starknet contract don't require a `main` function.
```
Comment (enitrat) [inline exercises/starknet/README.md:5]: Once again I don't really understand why this is part exists - it should be in the book directly

File: exercises/starknet/README.md [hunk 10]
```diff
@@ -1,7 +1,87 @@
...
+	// ContractAddress to Felt
+	let owner_felt: felt252 = contract_address_to_felt252(caller);
+	owner_felt.print();
+}
+```
 
 ## Further information
 
-- [Cross-contract interactions](https://cairo-book.github.io/ch09-00-00-abis-and-cross-contract-interactions.html)
+-   For more details, check out [The Structure of a Cairo Smart Contract](https://book.starknet.io/chapter_2/structure.html) in [the Starknet book](https://book.starknet.io).
```
Comment (enitrat) [inline exercises/starknet/README.md:87]: Why did you remove this link?

File: exercises/starknet/basics/starknet1.cairo [hunk 11]
```diff
@@ -1,28 +1,27 @@
 // starknet1.cairo
-// Address all the TODOs to make the tests pass!
-// Execute `starklings hint starknet1` or use the `hint` watch subcommand for a hint.
+// Starkling, Joe, is writing a really simple contract.
```
Comment (enitrat) [inline exercises/starknet/basics/starknet1.cairo:2]: I don't get why this file would be modified. The first approach was extremely simple and helped discover the following notions without too much complexity

- constructors & storage variable
- Modify contracts with external functions
- Read from contracts with view functions
- Learn how to create mappings

This new approach kinda does the same, but I find it a bit hard from starknet3 to starknet4

File: exercises/starknet/README.md (hunk 9 above)
Comment (shramee) [reply exercises/starknet/README.md:5]: Some basic info, to get started with...

File: exercises/starknet/README.md (hunk 10 above)
Comment (shramee) [reply exercises/starknet/README.md:87]: Aah, unintentional... I cleared everything we had in the beginning.
This is extremely important...

File: exercises/starknet/basics/starknet1.cairo (hunk 11 above)
Comment (shramee) [reply exercises/starknet/basics/starknet1.cairo:2]: But it was not really a contract :(

Comment (shramee) [reply exercises/starknet/basics/starknet1.cairo:2]: So the new flow is like,
1. Adding endpoints
2. Using storage (and ContractAddress intro)
3. and 4. Finally getting serious and writing more logic with authentication.

File: exercises/modules/modules1.cairo (hunk 1 above)
Comment (shramee) [reply exercises/modules/modules1.cairo:22]: Done

File: exercises/modules/readme.md (hunk 6 above)
Comment (shramee) [reply exercises/modules/readme.md:7]: Yeah, but Cairo book seemed a bit confusing to me.

File: exercises/modules/modules2.cairo (hunk 7 above)
Comment (shramee) [reply exercises/modules/modules2.cairo:29]: For this one they should figure it out themselves... From errors they get.

File: info.toml (hunk 3 above)
Comment (shramee) [reply info.toml:519]: Done

File: info.toml (hunk 4 above)
Comment (shramee) [reply info.toml:520]: They are already there so not sure `import` should be used..?
Refer could be used but is could be mildly confusing? Like, not sure if 'referring to say a function' is commonly used terminology?

File: exercises/modules/readme.md (hunk 6 above)
Comment (shramee) [reply exercises/modules/readme.md:7]: But it's there to refer to for more details...

File: exercises/starknet/README.md (hunk 10 above)
Comment (shramee) [reply exercises/starknet/README.md:87]: Fixed

**Comment by enitrat** [comment]
i'll try to solve them tonight and then approve

File: exercises/starknet/basics/starknet4.cairo [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    #[available_gas(2000000000)]
+    fn test_stock() {
+        let owner = util_felt_addr( 'Elizabeth' );
+        LizInventory::constructor(owner);
+
+        // Call contract as owner
+        starknet::testing::set_caller_address( owner );
+
+        // Add stock
+        LizInventory::__external::add_stock( util_span_2_items( 'Nano', 10 ) );
```
Comment (enitrat) [inline exercises/starknet/basics/starknet4.cairo:80]: add stock should be 
`fn add_stock(product_id: u8, new_stock: u128)`  or something like this

but not take a span as input, it makes thing complicated for nothing

File: exercises/starknet/basics/starknet4.cairo [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        let joe_score = util_get_span_first( LizInventory::__external::get_stock( util_span_1_item( 'Nano' ) ) );
```
Comment (enitrat) [inline exercises/starknet/basics/starknet4.cairo:81]: same for get stock. it should be `get_stock(product_id: u8)`

File: exercises/starknet/basics/starknet1.cairo [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+#[cfg(test)]
+mod test {
+    use array::ArrayTrait;
+    use array::SpanTrait;
+    use super::JoesContract;
+    #[test]
+    #[available_gas(2000000000)]
+    fn test_contract_view() {
+        JoesContract::__external::get_owner(ArrayTrait::new().span());
```
Comment (enitrat) [inline exercises/starknet/basics/starknet1.cairo:25]: Your tests should not use the function defined under the `__external` namespace! If you're unit testing, you have to call `JoesContract::get_owner`.
If you're testing against deployed contracts, you would call `joes_contract.get_owner()`

This should be modified in the following files as well.

File: exercises/starknet/basics/starknet4.cairo (hunk 12 above)
Comment (enitrat) [reply exercises/starknet/basics/starknet4.cairo:80]: ah I see, it's because you actually call it from the `__external` namespace, which we shouldn't do

File: info.toml [hunk 15]
```diff
@@ -502,3 +502,57 @@ If you're having trouble updating the distance value in the `Fish` and `Dog` imp
...
 3. Reconstruct `self` with the updated variables (`self = MyStruct { ... }`) 
 """
+
+# MODULES
+
+[[exercises]]
+name = "modules1"
+path = "exercises/modules/modules1.cairo"
+mode = "test"
+hint = """"""
```
Comment (enitrat) [inline info.toml:512]: If there are no hints, add "No hints this time!"

Comment (enitrat) [reply info.toml:512]: But I would add a hint here.
- You can bring a parent's modules items in the current module with `super::item_name`

File: exercises/starknet/basics/starknet4.cairo (hunk 12 above)
Comment (shramee) [reply exercises/starknet/basics/starknet4.cairo:80]: product id can be stringlike felt... New stock could be any `ux`

File: exercises/starknet/basics/starknet4.cairo (hunk 13 above)
Comment (shramee) [reply exercises/starknet/basics/starknet4.cairo:81]: `product_id` in many stores can be a string... Hence the felt is preferable...

File: exercises/starknet/basics/starknet1.cairo (hunk 14 above)
Comment (shramee) [reply exercises/starknet/basics/starknet1.cairo:25]: Starknet contract tests in `starkware-libs/cairo` repo seem to use `__external`...

File: exercises/starknet/basics/starknet4.cairo [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        // TODO: takes product_id and new_stock
+        // adds new_stock to stock in inventory
+        // only owner can call this
+    }
+
+    #[external]
+    fn purchase() {
+        // TODO: takes product_id and quantity
+        // subtracts quantity from stock in inventory
+        // anybody can call this, assert stock > quantity
```
[the remark this answers is not in the document]
Comment (shramee) [reply exercises/starknet/basics/starknet4.cairo:35]: Yeah, I saw that.
But I prefer explicit fail first pattern. So that it's obvious that the case is covered.

File: exercises/starknet/basics/starknet4.cairo (hunk 13 above)
Comment (enitrat) [reply exercises/starknet/basics/starknet4.cairo:81]: yeah sorry. I was talking about the signature, I thought you expected it to be a `Span<>`, that's what I first understood when reading the tests, since you're calling the functions from the __external namespace that expect serialized values.

you can forget about this

File: exercises/starknet/basics/starknet1.cairo (hunk 14 above)
Comment (enitrat) [reply exercises/starknet/basics/starknet1.cairo:25]: the result is the same and in our case it makes it _a lot_ simpler to read and understand

File: exercises/starknet/basics/starknet3.cairo [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        progress: LegacyMap<>
+    }
+
+    #[constructor]
+    fn constructor(owner: ContractAddress) {
+        contract_owner::write( owner );
+    }
+
+    #[external]
+    fn set_progress(user: ContractAddress, new_progress: felt252) {
```
[the remark this answers is not in the document]
Comment (shramee) [reply exercises/starknet/basics/starknet3.cairo:27]: Updated here... Couldn't find other places to update this.

File: exercises/starknet/basics/starknet4.cairo (hunk 13 above)
Comment (shramee) [reply exercises/starknet/basics/starknet4.cairo:81]: Updated to say `product` and specified `felt` in Storage comments.

File: exercises/starknet/basics/starknet1.cairo (hunk 14 above)
Comment (shramee) [reply exercises/starknet/basics/starknet1.cairo:25]: Yeah, but if down the road, someone unintentionally deletes/mistypes the attributes then tests would pass but...
So I thought using the `::external...` would be nice...

Comment (shramee) [reply exercises/starknet/basics/starknet1.cairo:25]: This is specially important for exercises where we are teaching them how to write contracts with working endpoints.
So a decided to go with this complication in tests.

File: info.toml (hunk 15 above)
Comment (shramee) [reply info.toml:512]: Sweet!

File: exercises/starknet/basics/starknet4.cairo (hunk 12 above)
Comment (shramee) [reply exercises/starknet/basics/starknet4.cairo:80]: Discussed in https://github.com/shramee/starklings-cairo1/pull/103#discussion_r1189186509

File: exercises/starknet/basics/starknet4.cairo (hunk 16 above)
[the remark this answers is not in the document]
Comment (shramee) [reply exercises/starknet/basics/starknet4.cairo]: Added comments to show explain our convo here...

**Comment by omarespejel** [comment]
Hey team! Unfortunately, I won't be participating in this PR, as I'm not yet fully familiar with how Starklings operates. However, I'll be getting up to speed this week so I can actively contribute to upcoming PRs. Thanks for understanding!

**Comment by shramee** [comment]
Hi Omar,
No worries, thanks for the update.

File: exercises/starknet/basics/starknet1.cairo (hunk 14 above)
Comment (enitrat) [reply exercises/starknet/basics/starknet1.cairo:25]: I'm strongly against it

Comment (shramee) [reply exercises/starknet/basics/starknet1.cairo:25]: But wouldn't the exercise pass even if they miss(type) the annotations. We don't want that.

You are welcome to improve them by writing deployment for class and testing that if you prefer. Maybe a helper util func that deploys and returns deployment to test?

**Comment by shramee** [comment]
@enitrat Tests updated

File: exercises/starknet/basics/starknet1.cairo (hunk 14 above)
Comment (shramee) [reply exercises/starknet/basics/starknet1.cairo]: > I'm strongly against it

Removed all `__external` calls.

**Review by enitrat — APPROVED** [review]
looks good :)
